src/query/models/tags.go

```diff
@@ -93,37 +88,118 @@ func (t Tags) IDLen() int {
 	return idLen
 }
 
-// IDWithExcludes returns a string representation of the tags excluding some tag keys.
-func (t Tags) IDWithExcludes(excludeKeys ...[]byte) uint64 {
-	b := make([]byte, 0, t.Len())
-	for _, tag := range t.Tags {
-		// Always exclude the metric name by default
-		if bytes.Equal(tag.Name, t.Opts.MetricName()) {
-			continue
+func (t Tags) quotedID() []byte {
+	needEscaping, length := t.escapingAndLength()
+	if needEscaping == nil {
+		return t.quoteIDSimple(length)
+	}
+
+	// TODO: pool these bytes.
+	id := make([]byte, length)
+	idx := 0
+	for i, tag := range t.Tags {
+		if needEscaping[i*2] {
+			idx = strconv.Escape(id, tag.Name, idx)
+		} else {
+			idx += copy(id[idx:], tag.Name)
 		}
 
-		found := false
-		for _, n := range excludeKeys {
-			if bytes.Equal(n, tag.Name) {
-				found = true
-				break
+		if needEscaping[i*2+1] {
+			idx = strconv.Quote(id, tag.Value, idx)
+		} else {
+			idx = strconv.QuoteSimple(id, tag.Value, idx)
```

Comment:
How much benefit does having `Quote` and `QuoteSimple` give you? Code would definitely be simpler without the distinction (though I could see how this would be faster).

Reply:
It's a really large difference;

```
10_tags,_2_len_easy-8         	 5000000	       328 ns/op	      64 B/op	       1 allocs/op
10_tags,_2_len_hard-8         	 3000000	       583 ns/op	     112 B/op	       2 allocs/op

100_tags,_2_len_easy-8        	  500000	      3207 ns/op	     704 B/op	       1 allocs/op
100_tags,_2_len_hard-8        	  300000	      5511 ns/op	    1104 B/op	       2 allocs/op

1000_tags,_2_len_easy-8       	   50000	     31613 ns/op	    8194 B/op	       1 allocs/op
1000_tags,_2_len_hard-8       	   30000	     59363 ns/op	   12292 B/op	       2 allocs/op

10_tags,_10_len_easy-8        	 3000000	       397 ns/op	     144 B/op	       1 allocs/op
10_tags,_10_len_hard-8        	 1000000	      1257 ns/op	     272 B/op	       2 allocs/op

100_tags,_10_len_easy-8     		  500000	      4007 ns/op	    1536 B/op	       1 allocs/op
100_tags,_10_len_hard-8       	  100000	     13479 ns/op	    2896 B/op	       2 allocs/op

1000_tags,_10_len_easy-8      	   30000	     41068 ns/op	   16388 B/op	       1 allocs/op
1000_tags,_10_len_hard-8      	   10000	    155339 ns/op	   29325 B/op	       2 allocs/op

10_tags,_100_len_easy-8       	 1000000	      1176 ns/op	    1152 B/op	       1 allocs/op
10_tags,_100_len_hard-8       	  200000	      9091 ns/op	    2080 B/op	       2 allocs/op

100_tags,_100_len_easy-8      	  200000	     10434 ns/op	   10880 B/op	       1 allocs/op
100_tags,_100_len_hard-8      	   20000	     90329 ns/op	   21969 B/op	       2 allocs/op

1000_tags,_100_len_easy-8     	   10000	    108446 ns/op	  106519 B/op	       1 allocs/op
1000_tags,_100_len_hard-8     	    2000	    908084 ns/op	  215157 B/op	       4 allocs/op

10_tags,_1000_len_easy-8      	  200000	      6704 ns/op	   10240 B/op	       1 allocs/op
10_tags,_1000_len_hard-8      	   20000	     89744 ns/op	   20512 B/op	       2 allocs/op

100_tags,_1000_len_easy-8     	   20000	     63663 ns/op	  106501 B/op	       1 allocs/op
100_tags,_1000_len_hard-8     	    2000	    850395 ns/op	  205065 B/op	       2 allocs/op

1000_tags,_1000_len_easy-8    	    2000	    650311 ns/op	 1008189 B/op	       3 allocs/op
1000_tags,_1000_len_hard-8    	     200	   8405176 ns/op	 2014826 B/op	      22 allocs/op
```

For most likely cases, there's a 3-10x increase, and this is just with quote characters; if using any long unicode this difference would get even worse.

Reply:
Ah ok, good to know. Which benchmark are these from? Do they compare preprocessing to determine which `Quote` method is needed followed by `Quote`/`QuoteSimple` to just running `Quote`?

Reply:
This is benchmarking the entire `ID()` method; `easy` has just valid characters, whereas `hard` has just `"`. The problem is that `utf8.DecodeRune` is quite a slow operation, which is called once per byte, and incurs a fair cost.